sigstore/_cli.py

```diff
@@ -47,343 +41,360 @@
-)
-@click.argument(
-    "files",
-    metavar="FILE [FILE ...]",
-    type=click.File("rb"),
-    nargs=-1,
-    required=True,
-)
-def _sign(
-    files: List[BinaryIO],
-    identity_token: Optional[str],
-    oidc_client_id: str,
-    oidc_client_secret: str,
-    oidc_issuer: str,
-    oidc_disable_ambient_providers: bool,
-    output_signature: Optional[str],
-    output_certificate: Optional[str],
-    fulcio_url: str,
-    rekor_url: str,
-    ctfe_pem: BinaryIO,
-    rekor_root_pubkey: BinaryIO,
-    staging: bool,
-) -> None:
-    # Fail if `--output-signature` or `--output-certificate` is specified with
-    # a value *and* we have more than one input. If passed without values,
-    # then treat them as an instruction to generate default {input}.sig and
-    # {input}.cert outputs for each {input}.
-    multiple_inputs = len(files) > 1
-    if (output_signature or output_certificate) and multiple_inputs:
-        click.echo(
+    # Stuff the parser back into our namespace, so that we can use it for
+    # error handling later.
+    args._parser = parser
+
+    if args.subcommand == "sign":
+        _sign(args)
+    elif args.subcommand == "verify":
+        _verify(args)
+    else:
+        parser.error(f"Unknown subcommand: {args.subcommand}")
+
+
+def _sign(args: argparse.Namespace) -> None:
+    # `--output` may not be mixed with `--output-signature` or `--output-certificate`.
+    if args.output and (args.output_signature or args.output_certificate):
+        args._parser.error(
+            "--output may not be combined with --output-signature or --output-certificate",
+        )
+
+    # Fail if `--output-signature` or `--output-certificate` is specified
+    # *and* we have more than one input.
+    if (args.output_signature or args.output_certificate) and len(args.files) > 1:
+        args._parser.error(
             "Error: --output-signature and --output-certificate can't be used with "
-            "explicit outputs for multiple inputs",
-            err=True,
+            "explicit outputs for multiple inputs; consider using --output",
         )
-        raise click.Abort
 
-    if staging:
+    # Build up the map of inputs -> outputs ahead of any signing operations,
+    # so that we can fail early if overwriting without `--force`.
+    output_map = {}
+    for file in args.files:
+        sig, cert = args.output_signature, args.output_certificate
+        if args.output:
+            sig = file.parent / f"{file.name}.sig"
+            cert = file.parent / f"{file.name}.crt"
+
+        if not args.force:
+            extant = None
+            if sig and sig.exists():
+                extant = sig
+            elif cert and cert.exists():
+                extant = cert
+
+            if extant:
+                args._parser.error(
+                    f"Refusing to overwrite pre-existing outputs without --force: {extant}"
+                )
```

Comment:
It might be a little annoying that this only tells the user about one of possibly two files that would be extant. I think this should output _all_ extant files at once.

Reply:
Yeah, agreed, will update.